### PR TITLE
Support postgresql+asyncpg schema for PostgresDsn

### LINKED ISSUE #2679: Can't use `postgresql+psycopg2://...` with `PostgresDsn `
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

I'm expecting to be able to use a database connection url such as `postgresql+psycopg2://user:pass@localhost:5432/app` (a format [SQLAlchemy is supporting](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#dialect-postgresql-psycopg2-connect)). 

However, it doesn't seem to accept the `+psycopg2` part. 

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.1
            pydantic compiled: False
                 install path: /Users/krisb/Code/fork/pydantic/pydantic
               python version: 3.7.10 (default, Apr 12 2021, 03:46:49)  [Clang 12.0.0 (clang-1200.0.32.29)]
                     platform: Darwin-20.3.0-x86_64-i386-64bit
     optional deps. installed: ['devtools', 'dotenv', 'email-validator', 'typing-extensions']
```

Changed test https://github.com/samuelcolvin/pydantic/blob/7af90a8d6c1c406799ad95b81e353b386cde5745/tests/test_networks.py#L318-L318 to include 
```py
assert Model(a='postgresql+psycopg2://user:pass@localhost:5432/app').a == 'postgresql+psycopg2://user:pass@localhost:5432/app'
```
pytest error output:
```
...
>           raise validation_error
E           pydantic.error_wrappers.ValidationError: 1 validation error for Model
E           a
E             URL scheme not permitted (type=value_error.url.scheme; allowed_schemes={'postgres', 'postgresql'})

pydantic/main.py:400: ValidationError
```

This is similar to https://github.com/samuelcolvin/pydantic/issues/1142, and a workaround for now is to use `AnyUrl` where this type of URL is supported.

### PR DESCRIPTION
**Pull request by fike**
## Change Summary

SQLAchemy 1.4 [added alpha support](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html) to async calls. The support to PostgreSQL needs [to add asyncpg](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html) in the schema, e.g.: `postgresql+asyncpg://scott:tiger@localhost/test`

## Related issue number

#2545
Closes #2679 

## Checklist

* [ X ] Unit tests for the changes exist
* [ X ] Tests pass on CI and coverage remains at 100%
* [ X ] Documentation reflects the changes where applicable
* [ X ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Review by PrettyWood — APPROVED** [review]
Thanks!

**Comment by Kludex** [comment]
As we are already here... Can we add `pg8000` as well? :sweat_smile:

**Comment by fike** [comment]
> As we are already here... Can we add `pg8000` as well? 😅

Certainly! I'm changing of the computer, if can waiting for a few days until finish my setup and implement it. :D

**Comment by fike** [comment]
> As we are already here... Can we add `pg8000` as well? 😅

Hi @Kludex, I added pg8000 here. :)

**Review by Kludex — APPROVED** [review]

File: docs/usage/types.md [hunk 1]
```diff
@@ -567,7 +567,7 @@ For URI/URL validation the following types are available:
 - `AnyUrl`: any scheme allowed, TLD not required
 - `AnyHttpUrl`: schema `http` or `https`, TLD not required
 - `HttpUrl`: schema `http` or `https`, TLD required, max length 2083
-- `PostgresDsn`: schema `postgres`, `postgresql` or `postgres+asyncpg`, user info required, TLD not required
+- `PostgresDsn`: schema `postgres`, `postgresql`, `postgres+asyncpg` or `postgresql+pg8000`, user info required, TLD not required
```
Comment (StephenBrown2) [inline docs/usage/types.md:570]:
```suggestion
- `PostgresDsn`: schema `postgres`, `postgresql`, `postgresql+asyncpg` or `postgresql+pg8000`, user info required, TLD not required
```

File: pydantic/networks.py [hunk 2]
```diff
@@ -298,7 +298,12 @@ class HttpUrl(AnyUrl):
 
 
 class PostgresDsn(AnyUrl):
-    allowed_schemes = {'postgres', 'postgresql', 'postgresql+asyncpg'}
+    allowed_schemes = {
+        'postgres',
+        'postgresql',
+        'postgresql+asyncpg',
+        'postgresql+pg8000'
+        }
```
Comment (StephenBrown2) [inline pydantic/networks.py]:
```suggestion
    }
```

File: docs/usage/types.md (hunk 1 above)
Comment (Kludex) [reply docs/usage/types.md]: It accepts both actually :sweat_smile:

File: changes/2567-postgres-asyncpg.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+* Added postgresql+asyncpg schema for PostgresDsn
+* Added postgresql+pg8000 schema for PostgresDsn
```
Comment (PrettyWood) [inline changes/2567-postgres-asyncpg.md:2]:
```suggestion
* Added postgresql+asyncpg scheme for PostgresDsn
* Added postgresql+pg8000 scheme for PostgresDsn
```

**Review by PrettyWood — CHANGES_REQUESTED** [review]
You can probably rebase on master too

**Comment by fike** [comment]
> You can probably rebase on master too

Hi @PrettyWood, can you check if I did right?  I'm not confident that if my laptop migration didn't mess something.

File: changes/2567-postgres-asyncpg.md [hunk 4]
```diff
@@ -0,0 +1,2 @@
+* Added postgresql+asyncpg scheme for PostgresDsn
+* Added postgresql+pg8000 scheme for PostgresDsn
```
Comment (PrettyWood) [inline changes/2567-postgres-asyncpg.md:2]:
```suggestion
Add `postgresql+asyncpg` and `postgresql+pg8000` schemes for `PostgresDsn`
```

**Review by PrettyWood — COMMENTED** [review]
Just a suggestion for the change file to use ` and present. LGTM otherwise for the patch :)

**Comment by fike** [comment]
> Just a suggestion for the change file to use ` and present. LGTM otherwise for the patch :)

Thanks! :)

**Review by PrettyWood — APPROVED** [review]

**Comment by fike** [comment]
> Hey @fike
> Seeing #2679, I wonder if we could just add all the dialects specified [here](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#dialect-postgresql)
> WDYT ?

Yes, certainly! I'll do that. 👍

File: docs/usage/types.md [hunk 5]
```diff
@@ -221,7 +221,7 @@ _(This script is complete, it should run "as is")_
     pydantic can't validate the values automatically for you because it would require
     consuming the infinite generator.
 
-#### Validating the first value
+## Validating the first value
```
Comment (PrettyWood) [inline docs/usage/types.md:224]:
```suggestion
#### Validating the first value
```
https://github.com/samuelcolvin/pydantic/pull/2577

File: docs/usage/types.md [hunk 6]
```diff
@@ -565,10 +565,16 @@ _(This script is complete, it should run "as is")_
 For URI/URL validation the following types are available:
 
 - `AnyUrl`: any scheme allowed, TLD not required
-- `AnyHttpUrl`: scheme `http` or `https`, TLD not required
-- `HttpUrl`: scheme `http` or `https`, TLD required, max length 2083
-- `PostgresDsn`: scheme `postgres`, `postgresql`, `postgresql+asyncpg` or `postgresql+pg8000`, user info required, TLD not required
-- `RedisDsn`: scheme `redis` or `rediss`, user info not required, tld not required (CHANGED: user info
+- `AnyHttpUrl`: schema `http` or `https`, TLD not required
+- `HttpUrl`: schema `http` or `https`, TLD required, max length 2083
+- `PostgresDsn`: scheme `postgres`, `postgresql`, user info required, TLD not required. Also, it's supported DBAPI dialects:
```
Comment (PrettyWood) [inline docs/usage/types.md:570]:
```suggestion
- `PostgresDsn`: scheme `postgres`, `postgresql`, user info required, TLD not required. Also, its supported DBAPI dialects:
```

File: docs/usage/types.md [hunk 7]
```diff
@@ -592,7 +598,7 @@ If you require a custom URI/URL type, it can be created in a similar way to the
 Assuming an input URL of `http://samuel:pass@example.com:8000/the/path/?query=here#fragment=is;this=bit`,
 the above types export the following properties:
 
-- `scheme`: always set - the url scheme (`http` above)
+- `scheme`: always set - the url schema (`http` above)
```
Comment (PrettyWood) [inline docs/usage/types.md:601]:
```suggestion
- `scheme`: always set - the url scheme (`http` above)
```

File: changes/2567-postgres-asyncpg.md [hunk 8]
```diff
@@ -1 +1 @@
-Add `postgresql+asyncpg` and `postgresql+pg8000` schemes for `PostgresDsn`
+Add `postgresql+asyncpg`, `postgresql+pg8000`, `postgresql+psycopg2cffi`, `postgresql+py-postgresql` and `postgresql+pygresql` schemes for `PostgresDsn`
```
Comment (PrettyWood) [inline changes/2567-postgres-asyncpg.md:1]: `postgresql+psycopg2` is missing ;)

**Review by PrettyWood — COMMENTED** [review]
Thanks!

File: changes/2567-postgres-asyncpg.md (hunk 8 above)
Comment (fike) [reply changes/2567-postgres-asyncpg.md]: Ouch! thanks! 👍

File: docs/usage/types.md [hunk 9]
```diff
@@ -565,10 +565,16 @@ _(This script is complete, it should run "as is")_
 For URI/URL validation the following types are available:
 
 - `AnyUrl`: any scheme allowed, TLD not required
-- `AnyHttpUrl`: scheme `http` or `https`, TLD not required
-- `HttpUrl`: scheme `http` or `https`, TLD required, max length 2083
-- `PostgresDsn`: scheme `postgres` or `postgresql`, user info required, TLD not required
-- `RedisDsn`: scheme `redis` or `rediss`, user info not required, tld not required (CHANGED: user info
+- `AnyHttpUrl`: schema `http` or `https`, TLD not required
+- `HttpUrl`: schema `http` or `https`, TLD required, max length 2083
```
Comment (samuelcolvin) [inline docs/usage/types.md:569]:
```suggestion
- `AnyHttpUrl`: scheme `http` or `https`, TLD not required
- `HttpUrl`: scheme `http` or `https`, TLD required, max length 2083
```

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM.

**Comment by samuelcolvin** [comment]
please update.

**Comment by fike** [comment]
> please update.

I did. Thanks, @samuelcolvin. :)

File: tests/test_networks.py [hunk 10]
```diff
@@ -19,6 +19,11 @@
         'postgres://user:pass@localhost:5432/app',
         'postgres://just-user@localhost:5432/app',
         'postgresql+psycopg2://postgres:postgres@localhost:5432/hatch',
+        'postgresql+asyncpg://user:pass@localhost:5432/app',
+        'postgresql+pg8000://user:pass@localhost:5432/app',
+        'postgresql+psycopg2cffi://user:pass@localhost:5432/app',
+        'postgresql+py-postgresql://user:pass@localhost:5432/app',
+        'postgresql+pygresql://user:pass@localhost:5432/app',
```
Comment (PrettyWood) [inline tests/test_networks.py:26]: All those tests would already be valid without your PR. You probably want to parametrize `test_postgres_dsns`

File: docs/usage/types.md [hunk 11]
```diff
@@ -567,8 +567,14 @@ For URI/URL validation the following types are available:
...
-- `PostgresDsn`: scheme `postgres` or `postgresql`, user info required, TLD not required
-- `RedisDsn`: scheme `redis` or `rediss`, user info not required, tld not required (CHANGED: user info
+- `PostgresDsn`: scheme `postgres`, `postgresql`, user info required, TLD not required. Also, its supported DBAPI dialects:
+  - `postgresql+psycopg2`
+  - `postgresql+asyncpg`
+  - `postgresql+pg8000`
+  - `postgresql+psycopg2cffi`
+  - `postgresql+py-postgresql`
+  - `postgresql+pygresql`
+- `RedisDsn`: schema `redis` or `rediss`, user info not required, tld not required (CHANGED: user info
```
Comment (PrettyWood) [inline docs/usage/types.md:577]:
```suggestion
- `RedisDsn`: scheme `redis` or `rediss`, user info not required, tld not required (CHANGED: user info
```

**Review by PrettyWood — CHANGES_REQUESTED** [review]

**Comment by PrettyWood** [comment]
Please update and rebase/merge with master for CI to run properly. Sorry for reaching out so late. Thanks!

File: docs/usage/types.md (hunk 11 above)
Comment (fike) [reply docs/usage/types.md]: Right, done!

File: tests/test_networks.py (hunk 10 above)
Comment (fike) [reply tests/test_networks.py:26]: I did that but I'm sure if my implementation was good. It's wonderful anyone can review it. :)

**Comment by fike** [comment]
@PrettyWood, can you check if I rebased correctly? :)

**Review by PrettyWood — APPROVED** [review]
Thanks!

**Pull request merged**